File: lab-03.hs
module Lab3 where

-----------------------------------------------------------------------------------------------------------------------------
-- LIST COMPREHENSIONS
------------------------------------------------------------------------------------------------------------------------------

-- ===================================
-- Ex. 0 - 2
-- ===================================

evens :: [Integer] -> [Integer]
evens = \xs -> [x | x <- xs, x `mod` 2 == 0]

-- ===================================
-- Ex. 3 - 4 
-- ===================================

-- complete the following line with the correct type signature for this function
squares :: Integer -> [Integer]
squares n = [x * x | x <- [1..n]]

sumSquares :: Integer -> Integer
sumSquares n = sum (squares n)

-- ===================================
-- Ex. 5 - 7
-- ===================================

-- complete the following line with the correct type signature for this function
squares' :: Integer -> Integer -> [Integer]
squares' m n = [x * x | x <- [i..j]]
  where
    i = n + 1
    j = m + n

sumSquares' :: Integer -> Integer
sumSquares' x = sum . uncurry squares' $ (x, x)

-- ===================================
-- Ex. 8
-- ===================================

coords :: Integer -> Integer -> [(Integer,Integer)]
coords n m = [ (x,y) | x <- [0..n], y <- [0..m] ]

File: lab-03.scala
val evens: List[Int] => List[Int] = 
  xs => for { x <- xs.filter(_ % 2 == 0) } yield x

val squares: Int => Seq[Int] = 
  n => for { x <- 1 to n } yield x * x

val sumSquares: Int => Int =
  n => squares(n).foldLeft(0)(_+_)

val _squares: Int => Int => Seq[Int] =
  m => {
    n => {
      val i  = n + 1
      val j  = m + n
      for {
        x <- i to j
      } yield x * x
    }
  }
  
val _sumSquares: Int => Int =
  x => Function.uncurried[Int,Int,Seq[Int]](_squares)(x, x).foldLeft(0)(_+_)

val coords: Int => Int => Seq[(Int,Int)] = 
  n => {
    m => for { 
      x <- 0 to n
      y <- 0 to m
      } yield (x , y)
  }
